### ISSUE TITLE
Proposal: Native Packages, enabling Supervisor service support for platforms without packages.

### ISSUE THREAD

**Issue by atrniv** [habitat-sh/habitat#8501]
## Summary

This is a proposal for introducing the idea of habitat package types. There will primarily be two types of packages:
- **Standard Packages** : These are currently the only type of packages we have. A standard package is built in a cleanroom environment and it is guaranteed to run in any environment once it has been built. This is possible because every dependency of a standard package is also another standard package.
- **Native Packages**: This is a new type of package. Native packages are built in the host environment that the habitat packaging tool gets run in. These packages transfer the responsibility of managing runtime dependencies to the user. A native package is essentially a thin wrapper around an existing application that makes use of hooks to define how to manage the application lifecycle.

## Motivation

The primary advantages we gain by introducing the concept of a *Native Package* is the ability to run the Habitat Supervisor on platforms where we currently do not have a package ecosystem bootstrapped (eg: ARM, Solaris, M1 Macs). This will enable us to open a broader set of use cases for Habitat usage.

## Design

To enable support for native packages, the `HAB_NATIVE_PACKAGE_SUPPORT` environment variable must be set to `true`.

To create a Native Package you must:
- Put `# package_type: "native"` as the topmost line of your `plan.sh` / `plan.ps1` file.
- You cannot have the following plan parameters:
  - pkg_deps
  - pkg_build_deps
  - pkg_interpreters 
- You must have a run hook defined for your package. 
- You can use any interpreter for your hooks. You must use a shebang line at the start of all your hook files with an appropriate path to the interpreter you wish to use.
- There will be a new Metafile in a package called `PKG_TYPE`. It will contain `native` or `standard` to indicate the type of package.

## Behavior Changes
- Invoking `hab pkg build` will automatically run the hab plan build script on the host system when it detects that the plan builds a native package.
- Invoking `build` within a studio environment for a native package plan will result in an error message suggesting the user to run `hab pkg build` outside the studio.
- Adding any unsupported plan parameter such as `pkg_deps` will result in a build time error message.
- Native Package hooks will have to ensure that the interpreter they use and any commands that are used in them are available in the runtime environment.

## Design Issues Considered

- Supporting dependencies between native packages is likely to be error prone as the user has to ensure that they always build and run the package in an environment that is compatible with all dependencies. Hence we will also not be supporting native packages that are libraries at this time.
- We need to introduce a new concept here (ie: Package Type) in order to have a clean way to detect and prevent mixing up of packages meant to work in different environments.
- There is no way to have both a native and a standard package plan file in the same folder, due to the existing rules of writing plan files.

## Things yet to be decided

- How will the builder APIs have to be modified to support this.

**Comment by mwrock** [comment]
Instead of `HAB_NATIVE_PACKAGE_SUPPORT ` we should use `HAB_FEAT_NATIVE_PACKAGE_SUPPORT` and leverage our existing feature flag capabilities.

Another thing to highlight for motivation is the desire to separate the concerns of packaging/dep management from Habitat runtime environment. Today, in order to use a habitat supervisor, you need to go "all in" on packaging all of your applications dependencies all the way down to glibc which can be extremely burdensome on a new platform. We would like to shorten the onramp for users interested in the supervisor and basic packaging but only plan to build and run their apps on a single platform.

**Comment by gscho** [comment]
A few questions:

> # package_type: "native"
Why a magic comment and not a `pkg_type=native`  bash/ps variable following the existing pattern?

> There will be a new Metafile in a package called PKG_TYPE
Why is this needed if there is a `pkg_type` setting in the plan file?

**Comment by atrniv** [comment]
@mwrock , I've elaborated on the motivation to hopefully make it more clearer and modified the feature flag to enable this as per your suggestions.

@gscho , I've added an entry in the design consideration section on why a magic comment is a better option to a bash/ps variable for the purpose of determining the plan type. 

Regarding the `PKG_TYPE` metafile, this is a file that is automatically created by Habitat during the build process and included within the package's `.hart` file. It is meant to convey the package type information to other packages that may try to depend on it during the habitat build process to enforce the build restrictions outlined in the proposal. You do not need to actually create this file anywhere in your application source tree.

**Comment by atrniv** [comment]
I'm having seconds thoughts about the `# package_type: native` magic comment approach as it prevents us from having native and standard package plans in the same folders. 

We have such a use case within habitat itself with the `hab-plan-build` package. When doing a native build on the host system we currently don't want any kind of interaction with builder for the initial prototype of this feature. 
One way I could achieve this is by embedding the scripts into the binary and then writing them out to a temporary file and running the script from that on the host system to do the plan build.

However in the future when we support native packages on builder we may want to instead introduce a proper `hab-plan-build-native` package and execute it instead. In that future situation it would be hard to re-use the same hab-plan-build source as it can only contain one `plan.sh` file which is either a native or standard package based on the current design proposal.

Another slightly less ergonomic (in my opinion) alternative that might solve this problem are by introducing the following rules instead:
- Native plans must be written in a `native-plan.sh` file. They can sit along side a regular `plan.sh` file.
- To build native packages you must run `hab pkg build --native-package`.

@mwrock, @gscho any thoughts on the same?

**Comment by atrniv** [comment]
@mwrock , the more I've been working on this, the less of a difference I'm seeing between native habitat packages and standard habitat packages. 

There are also valid use cases for standard packages depending on native packages in the [bootstrapping process plans](https://github.com/habitat-sh/bootstrap-plans). For example, any of the build-tools packages which are built before the standard `core/glibc` is built will have a runtime dependency on `core/build-tools-glibc` which is a native package because it's build process is not isolated. However it is absolutely portable to any `aarch64-unknown-linux` based linux distribution with a compatible kernel like the standard `core/glibc` package.

The runtime isolation of a package has to be achieved through custom modification of the build process for each package. The studio does not directly affect runtime isolation, it can only guarantee build time isolation. Even if you do an isolated build of a package, it can always a library / dependency from the host environment through various mechanisms like `LD_LIBRARY_PATH` or just hardcoded file paths. 

Habitat cannot prevent this runtime behavior because the supervisor does not run the process in a isolated environment (unless you export to a docker container). So the portability of a package has very little to do with the build time isolation of a package, it depends on the modifications done to ensure runtime isolation and the package's source code itself.

The portability of a package plan however can be guaranteed through build time isolation. This does not mean however that the resulting package is actually portable and has runtime isolation.

In the light of these points, I believe the way that we decide a package is a native package based on a non-isolated build right now may be wrong.

If standard packages means runtime portability, then the only person who can tell us if the package is a native package or a standard package is the plan author regardless of the build process. He could do this through a simple `pkg_type` variable in the plan.

**Comment by mwrock** [comment]
Yeah I think I agree. The more I think of this, the less I like the name "native" package because it hints that there is something special about these packages and there really isn't. The only thing really different here is we are providing a way to build a package outside of a studio.

The big concern here is that if it is easy to upload packages built outside of a studio and a popular core package is built with a dependency to such a package, then it is possible the top level package may fail at run time in an unknown but supported environment. The same concern exists if the top level package itself is built outside of a studio, which would have an even higher potential for danger.

You mention that the studio does not really provide runtime guarantees which is true but this certainly increases the risk of missing deps quite a bit.

That said, the only way a native package could be added as a dep to a core package would be if it were uploaded by a member with maintainer priviliges to the origin we designate for these packages. That should only exist for internal habitat employees. Assuming we "know what we are doing" then I think this is acceptable.

**Comment by mwrock** [comment]
closed via #8547
